### PR TITLE
Alter table freeze with verbose output

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce setting `alter_partition_verbose_result` which outputs information about touched parts for some types of `ALTER TABLE ... PARTITION ...` queries (currently `ATTACH` and `FREEZE`). Closes #8076.

### CODE REVIEW

File: src/Interpreters/InterpreterAlterQuery.cpp [hunk 1]
```diff
@@ -86,7 +87,13 @@ BlockIO InterpreterAlterQuery::execute()
     if (!partition_commands.empty())
     {
         table->checkAlterPartitionIsPossible(partition_commands, metadata_snapshot, context.getSettingsRef());
-        table->alterPartition(query_ptr, metadata_snapshot, partition_commands, context);
+        auto partition_commands_pipes = table->alterPartition(query_ptr, metadata_snapshot, partition_commands, context);
+        if (!partition_commands_pipes.empty())
+        {
+            QueryPipeline pipeline;
+            pipeline.init(std::move(partition_commands_pipes));
+            res.pipeline = std::move(pipeline);
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterAlterQuery.cpp:95]: Just `res.pipeline.init(std::move(partition_commands_pipes))`

File: src/Storages/PartitionCommands.h [hunk 2]
```diff
@@ -66,9 +67,39 @@ struct PartitionCommand
...
+    String backup_name;
+};
+
+using PartitionCommandsResultInfo = std::vector<PartitionCommandResultInfo>;
+
+/// Convert partition comands result to Source from single Chunk, which will be
+/// used to print info to the user. Tries to create narrowest table for given
+/// results. For example, if all commands were FREEZE commands, than
+/// old_part_name column will be absent.
+std::shared_ptr<SourceFromSingleChunk> convertCommandsResultToSource(const PartitionCommandsResultInfo & commands_result);
```
Comment (KochetovNicolai) [inline src/Storages/PartitionCommands.h:103]: Probably, we can already return `Pipes` here. No need to specify concrete source type.
(If you really want, declare `SourceFromSingleChunk` and remove `#include <Processors/Sources/SourceFromSingleChunk.h>` from header)

File: src/Storages/PartitionCommands.cpp [hunk 3]
```diff
@@ -97,4 +99,86 @@ std::optional<PartitionCommand> PartitionCommand::parse(const ASTAlterCommand *
         return {};
 }
 
+std::string PartitionCommand::typeToString() const
```
Comment (KochetovNicolai) [inline src/Storages/PartitionCommands.cpp:102]: Maybe we can use this method later in `ASTAlterCommand::formatImpl`

**Review by KochetovNicolai — APPROVED** [review]

File: src/Storages/PartitionCommands.cpp (hunk 3 above)
Comment (alesapin) [reply src/Storages/PartitionCommands.cpp:102]: Probably we can, but it require a lot of changes and I'm not sure that code will became much simplier.

**Comment by alesapin** [comment]
Tests OK.

**Comment by nvartolomei** [comment]
Why return relative path instead of absolute? If multiple disks/volumes are used the path is ambiguous if I understand correctly.

**Pull request merged**